### PR TITLE
feat(platform): allow new tokens on contract update and refactor contract struct validations

### PR DESCRIPTION
**Pull request by pauldelucia**
## Issue being fixed or feature implemented

Allow new tokens on contract update, add advanced structure validation for contract update, basic structure validation for contract update and create.

## What was done?

Remove the check for new tokens and replace with validations. Refactor contract validations.

## How Has This Been Tested?

Added unit tests

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

## Summary by CodeRabbit

- **New Features**
  - Introduced basic structure validation for data contract creation and update transitions, including checks for group and token configuration integrity.
  - Added advanced structure validation versioning for data contract creation transitions, supporting both initial and updated validation logic.
  - Enhanced token validation to allow new tokens with additional checks on distribution timestamps.
  - Added version-dependent conditional basic structure validation for data contract create and update state transitions.

- **Bug Fixes**
  - Improved error reporting for invalid token configurations, such as non-contiguous positions, excessive base supply, and missing localization.
  - Updated error handling to classify certain validation errors as unpaid consensus errors.

- **Tests**
  - Expanded test coverage for token and group validation scenarios, including edge cases and error conditions.
  - Added tests for token addition with gaps, large base supply, and invalid localization.
  - Added tests verifying token addition with valid and invalid pre-programmed distribution timestamps.

- **Chores**
  - Updated platform and validation versioning to support new validation logic and maintain compatibility with previous versions.
  - Refined method signatures to include block information for enhanced validation context.
  - Simplified method signatures and validation flows by removing redundant parameters and delegating to newer validation versions.

### CODE REVIEW

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_create/basic_structure/v0/mod.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+            if !validation_result.is_valid() {
+                return Ok(SimpleConsensusValidationResult::new_with_error(
+                    validation_result
+                        .errors
+                        .first()
+                        .cloned()
+                        .expect("error should exist"),
+                ));
+            }
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_create/basic_structure/v0/mod.rs:49]: Just do ```            if !validation_result.is_valid() {
                return Ok(validation_result);
            }```

Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_create/basic_structure/v0/mod.rs:49]: and same for others here.

File: packages/rs-drive-abci/src/execution/validation/state_transition/processor/v0/mod.rs [hunk 2]
```diff
@@ -532,15 +530,16 @@ impl StateTransitionBasicStructureValidationV0 for StateTransition {
...
         }
     }
     fn has_basic_structure_validation(&self) -> bool {
-        !matches!(
-            self,
-            StateTransition::DataContractCreate(_)
-                | StateTransition::DataContractUpdate(_)
-                | StateTransition::MasternodeVote(_)
-        )
+        !matches!(self, StateTransition::MasternodeVote(_))
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/validation/state_transition/processor/v0/mod.rs:542]: This won't work, because it's dependent on the platform version. I think we should check the platform version instead to check to see if it has basic structure validation for the version we currently are on.

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_create/mod.rs [hunk 3]
```diff
@@ -72,6 +76,32 @@ impl StateTransitionActionTransformerV0 for DataContractCreateTransition {
...
+            Some(0) => self.validate_basic_structure_v0(platform_version),
+            Some(version) => Err(Error::Execution(ExecutionError::UnknownVersionMismatch {
+                method: "data contract create transition: validate_basic_structure".to_string(),
+                known_versions: vec![0],
+                received: version,
+            })),
+            None => Err(Error::Execution(ExecutionError::VersionNotActive {
+                method: "data contract create transition: validate_basic_structure".to_string(),
+                known_versions: vec![0],
+            })),
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_create/mod.rs:100]: In this case we might just want to return Ok(Valid)

Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_create/mod.rs:100]: Actually probably not, we just don't want to be able to get here.

File: packages/rs-platform-version/src/version/drive_abci_versions/drive_abci_validation_versions/v1.rs [hunk 4]
```diff
@@ -80,7 +80,7 @@ pub const DRIVE_ABCI_VALIDATION_VERSIONS_V1: DriveAbciValidationVersions =
                 transform_into_action: 0,
             },
             contract_create_state_transition: DriveAbciStateTransitionValidationVersion {
-                basic_structure: None,
+                basic_structure: Some(0),
```
Comment (QuantumExplorer) [inline packages/rs-platform-version/src/version/drive_abci_versions/drive_abci_validation_versions/v1.rs:83]: This can't change.

File: packages/rs-platform-version/src/version/drive_abci_versions/drive_abci_validation_versions/v1.rs [hunk 5]
```diff
@@ -89,7 +89,7 @@ pub const DRIVE_ABCI_VALIDATION_VERSIONS_V1: DriveAbciValidationVersions =
                 transform_into_action: 0,
             },
             contract_update_state_transition: DriveAbciStateTransitionValidationVersion {
-                basic_structure: None,
+                basic_structure: Some(0),
```
Comment (QuantumExplorer) [inline packages/rs-platform-version/src/version/drive_abci_versions/drive_abci_validation_versions/v1.rs:92]: This can't change.

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

**Review by QuantumExplorer — APPROVED** [review]

**Review by QuantumExplorer — APPROVED** [review]
